### PR TITLE
Fix part of #5: Abstraction on top of Oppia GAE for Topic Management System

### PR DESCRIPTION
**Pull request by rt4914**
## Explanation
The PR contains data module, necessary libraries for API calls and Topic Management System models and retrofit instance.

## Libraries Used
1. **Retrofit:** implementation 'com.squareup.retrofit2:retrofit:2.5.0'
This library is responsible for making API calls to Oppia backend
2. **Retrofit GSON Convertor:** implementation 'com.squareup.retrofit2:converter-gson:2.5.0'
This library is a temporary fix for XSSI_PREFIX. This library can be removed once we use interceptor which will remove the XSSI_PREFIX from the response.
3. **Retrofit Moshi Convertor:** implementation 'com.squareup.retrofit2:converter-moshi:2.5.0'
This library is the actual deserializer that we are using over GSON.
4. **Moshi Kotlin Support:** implementation"com.squareup.moshi:moshi-kotlin:1.8.0"
Moshi Kotlin support library for using Moshi in Kotlin.

## Issues
The models in this PR does not have custom field names because @Json in Moshi which is similar to @SerializedName in GSON is not working with Kotlin Moshi.

## Checklist
Work items:

- [x] Identify among all existing GAE endpoints (get/post), which we can use during the prototype
- [x] Loosely identify data missing from the above endpoints (this bucket is expected to be small/non-existent), and identify which endpoints have data which needs to be changed for the Android app
- [x] Specify the Retrofit interfaces to represent all of the identified endpoints (simple interfaces/no annotations or return types needed beyond pseudocode)
- [x] Identify** what data (via pseudocode data structures/YAML/JSON/etc) we will be passing along from the backend through these endpoints
- [ ] Identify how error handling will work (e.g. what Retrofit does), whether RPC retry is supported & how it is/can be configured, etc.
- [ ] Determine the testing strategy for the GAE endpoints for downstream app components

### CODE REVIEW

**Comment by rt4914** [comment]
@BenHenning I have removed Retrofit-Gson-Converter library and made introduces an interceptor to modify the json response to remove XSSI_PREFIX. Now everything is working perfectly without using any extra libraries.

**Comment by rt4914** [comment]
@BenHenning PTAL

File: .idea/dictionaries/talesra.xml
Comment (seanlip) [inline .idea/dictionaries/talesra.xml]: Please add newline before EOF, here and elsewhere.

You can check the "Files Changed" tab to see which files need it.

File: .gitignore
Comment (seanlip) [inline .gitignore]: I don't think we should have per-user details in .gitignore. @BenHenning?

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (seanlip) [inline data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: You might as well do this now, here and elsewhere? Try to avoid leaving TODOs behind you in the codebase.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (seanlip) [inline data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: All TODOs should be formatted as follows:

``` TODO(#...): ...```

where the thing in parentheses is an issue number that describes what needs to be done in detail.

File: .idea/dictionaries/talesra.xml
Comment (rt4914) [reply .idea/dictionaries/talesra.xml]: @seanlip I have added new line at the end in the files which I have written and some files are automatically generated in which EOF is not there. So should I change those generated files too? Like the one which you have mentioned.

Comment (seanlip) [reply .idea/dictionaries/talesra.xml]: Can you find a way to generate them so that a newline is added at EOF?

Comment (seanlip) [reply .idea/dictionaries/talesra.xml]: (In any case, the short answer to your question is yes. Though it may be worth checking with @BenHenning whether we should be checking in generated files into source control.)

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Actually these TODOs are just a small part of a larger PR that we will have to do at later stage. Thus cannot solve this right now.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: These TODOs do not have an issue number yet. Should I create an issue and than change the TODO accordingly?

File: .idea/dictionaries/talesra.xml
Comment (rt4914) [reply .idea/dictionaries/talesra.xml]: @seanlip I have activated this from settings in android studio.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (seanlip) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Yes, please.

File: .idea/dictionaries/talesra.xml
Comment (seanlip) [reply .idea/dictionaries/talesra.xml]: I don't understand, sorry. What have you activated?

Also, please check your spelling in .gitignore. That might be source of some of the issues.

Comment (rt4914) [reply .idea/dictionaries/talesra.xml]: > Can you find a way to generate them so that a newline is added at EOF?

In Android Studio -> Settings -> Editor Settings we have an option to select "Add newline before EOF" in generated files. That way even the generated files will have a new line before EOF.

Comment (rt4914) [reply .idea/dictionaries/talesra.xml]: This has been solved.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: 2 TODOs removed and made changes to remaining TODO

**Comment by rt4914** [comment]
Hi @seanlip , I generally put extra lines in between code sections for more clear code vision (Check Topic.kt file for the that), so should I continue doing that or not?

File: .idea/dictionaries/talesra.xml
Comment (BenHenning) [inline .idea/dictionaries/talesra.xml]: Since this is specifically for you, maybe add it to .gitignore or, if possible, add these words to a shared dictionary so that everyone can benefit from them.

**Review by BenHenning — CHANGES_REQUESTED** [review]
My initial comment overlaps with one of Sean's earlier comments.

However, before I do a full pass on this can you rebase this & #62 on top of the introduce-domain-module branch and change the PR's base branches to introduce-domain-module? That way you can remove the new module bits and we can review just the new code in the context of what #64 is checking in.

File: .idea/dictionaries/talesra.xml
Comment (rt4914) [reply .idea/dictionaries/talesra.xml]: Added shared project dictionary. Though currently IDE only add words to the <username> based dictionary, therefore developers will have to add words manually in this.

**Comment by rt4914** [comment]
> My initial comment overlaps with one of Sean's earlier comments.
> 
> However, before I do a full pass on this can you rebase this & #62 on top of the introduce-domain-module branch and change the PR's base branches to introduce-domain-module? That way you can remove the new module bits and we can review just the new code in the context of what #64 is checking in.

Hi Ben,
I have done the rebase and changed the PR's base branch for this PR only. Actually, I am using rebase for the first time and therefore I just want to confirm if its correct or not.

Also, if this PR is correct then probably you should just check my #69 PR because that contains all the API services and models and therefore finally if #69 PR is finalised then we can close other PRs for GAE abstraction.

I have checked all the Models and services with valid API calls, so they are working as intended.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Please expand imports instead of using wildcards. It helps avoid namespace pollution, and provides better context to readers on needed dependencies.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: For here & all other new classes in this PR: please add Javadoc documentation for them.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Please move ``200`` to a module constant to provide context to it so that it's not a seemingly magic number. If Retrofit has a constant for it, that would be even better.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Should we maybe just remove the XSSI_PREFIX instead? Extra whitespace should just be ignored by the JSON parser, and this guards against the assumption that a newline needs to be present for the prefix.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Per our conversation last week, is there action needed to be taken here? I thought we were just going to rely on the default Retrofit behavior for error handling (e.g. that errors should be propagated to the Retrofit Call objects.

File: data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt:7]: Please document this & next comment.

Also, please add a TODO on #74 to move these to the DI graph.

File: data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt]: Nit: need space before ``{``

Ditto for other files in this PR.

File: data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt]: Rather than having a var store retrofit, could we use lazy initialization and keep the field as a value?

See: https://www.baeldung.com/kotlin-lazy-initialization.

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt]: Let's just omit this comment. If we aren't including these, I don't think we need to include them.

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt]: Please frame this as a Javadoc instead of a line comment. Ditto elsewhere for similar comments.

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt]: For lists & collections of objects, can they actually be absent?

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:10]: Please add Javadocs for these data objects (here & elsewhere). If possible, it'd be really useful to include links to the Python code that these structures correspond to for reference.

File: data/src/main/java/org/oppia/data/backends/gae/model/SubtopicSummary.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/model/SubtopicSummary.kt]: Perhaps ``subtopicId`` might be clearer?

File: data/src/main/java/org/oppia/data/backends/gae/model/CanonicalStorySummary.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/model/CanonicalStorySummary.kt]: Perhaps ``storyId`` instead?

File: data/src/main/java/org/oppia/data/backends/gae/model/AdditionalStorySummary.kt
Comment (BenHenning) [inline data/src/main/java/org/oppia/data/backends/gae/model/AdditionalStorySummary.kt]: Are there any differences between this & the canonical version? Are they also different structures in the backend? It might be nice to consolidate them into 1 object if not, since they seem to be the same structure with different uses.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @rt4914! There should be a PR coming out soon for the final module to wrap up  #4. We may block this until that PR is out to avoid introducing the data module twice.

**Comment by BenHenning** [comment]
Actually one more request: @rt4914 can you also add tests for the interceptor? Since it's doing logic, we should have tests to ensure it continues to do the right thing in the long-term, if possible.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Solved this

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Added Javadoc to every class and also changed ModifyJsonInterceptor to NetworkInterceptor

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Created new Constants file at module level and introduced 200 as a constant.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Code updated and used removePrefix method.

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt]: Yes, this is not required now and has been removed from the code.

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt]: I have seen responses in which lists and collections are also absent as well as null.

File: data/src/main/java/org/oppia/data/backends/gae/model/SubtopicSummary.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/model/SubtopicSummary.kt]: Changed id to subtopicId

File: data/src/main/java/org/oppia/data/backends/gae/model/CanonicalStorySummary.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/model/CanonicalStorySummary.kt]: Changed id to storyId

File: data/src/main/java/org/oppia/data/backends/gae/model/AdditionalStorySummary.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/model/AdditionalStorySummary.kt]: Merged Additional and Canonical into "StorySummary" model.

File: data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt]: Solved

File: data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/NetworkSettings.kt:7]: I have added a TODO comment. Not sure if you wanted me to do something else. So please check this once again.

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt:10]: Added Javadoc comment and python file links

File: .gitignore
Comment (rt4914) [reply .gitignore]: Solved from code side. Removed talesra.xml and introduced a common project-level dictionary

File: data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/OppiaGaeClient.kt]: Solved. Made use of kotlin-lazy as per Ben's suggestion.

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt]: Removed this comment.

File: data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt
Comment (rt4914) [reply data/src/main/java/org/oppia/data/backends/gae/model/Topic.kt]: Changed comments to Javadoc format

**Comment by rt4914** [comment]
> Actually one more request: @rt4914 can you also add tests for the interceptor? Since it's doing logic, we should have tests to ensure it continues to do the right thing in the long-term, if possible.

Hi @BenHenning , I have written test cases for NetworkInterceptor. I have taken help from your domain testing code and https://developer.android.com/training/testing/unit-testing/local-unit-tests

As this is me first time experience in writing a test cases please have a good look. Also, I have checked and both of these test cases pass.

**Comment by seanlip** [comment]
Thanks @rt4914 -- no further comments from my end. I'll need to defer the rest of the review to @BenHenning.

**Comment by rt4914** [comment]
@BenHenning , I have created another PR for this on #75 because there were a lot of merge conflicts on this PR. I will close this PR once you merge #75 PR.  Also, just for clarification the new branch name is gae-abstraction-topic-page-2

**Comment by BenHenning** [comment]
Closing this since it's been replaced by #78.

**Pull request closed without merging**
